crates/typst/src/layout/grid/layout.rs

```diff
@@ -216,38 +226,123 @@ impl CellGrid {
     ) -> SourceResult<Self> {
         // Number of content columns: Always at least one.
         let c = tracks.x.len().max(1);
 
-        // If not all columns in the last row have cells, we will add empty
-        // cells and complete the row so that those positions are susceptible
-        // to show rules and receive grid styling.
-        // We apply '% c' twice so that 'cells_remaining' is zero when
-        // the last row is already filled (then 'cell_count % c' would be zero).
+        // We can't just use the cell's index in the 'cells' vector to
+        // determine its automatic position, since cells could have arbitrary
+        // positions, so the position of a cell in 'cells' can differ from its
+        // final position in 'resolved_cells' (see below).
+        // Therefore, we use a counter, 'auto_index', to determine the position
+        // of the next cell with (x: auto, y: auto). It is only stepped when
+        // a cell with (x: auto, y: auto), usually the vast majority, is found.
+        let mut auto_index = 0;
+
+        // We have to rebuild the grid to account for arbitrary positions.
+        // Create at least 'cells.len()' positions, since there will be at
+        // least 'cells.len()' cells, even though some of them might be placed
+        // in arbitrary positions and thus cause the grid to expand.
         let cell_count = cells.len();
-        let cells_remaining = (c - cell_count % c) % c;
-        let cells = cells
-            .iter()
-            .cloned()
-            .chain(std::iter::repeat_with(T::default).take(cells_remaining))
+        let mut resolved_cells: Vec<Option<Cell>> = Vec::with_capacity(cell_count);
+        for cell in cells.iter().cloned() {
+            // Let's calculate the cell's final position based on its
+            // requested position.
+            let resolved_index = {
+                let cell_x = cell.x(styles);
+                let cell_y = cell.y(styles);
+                resolve_cell_position(cell_x, cell_y, &resolved_cells, &mut auto_index, c)
+                    .at(span)?
+            };
+            let x = resolved_index % c;
+            let y = resolved_index / c;
+
+            // Let's resolve the cell so it can determine its own fields
+            // based on its final position.
+            let cell = cell.resolve_cell(
+                x,
+                y,
+                &fill.resolve(engine, x, y)?,
+                align.resolve(engine, x, y)?,
+                inset,
+                styles,
+            );
+
+            // Now let's check if the cell's position is valid.
+            if let Some(existing_cell) = resolved_cells.get_mut(resolved_index) {
+                // We are trying to place a cell in an existing position.
+                // Ensure we aren't trying to place a cell where there is
+                // already one.
+                if existing_cell.is_some() {
+                    bail!(
+                        span,
```

Comment:
should we really fail at the generic span here or at the span of the second cell?

Reply:
I think it can be fine to use the span of the second cell, I was mostly just afraid of the span being in some totally unrelated location, leading the error to be potentially confusing. (Could be worse if e.g. you're doing something like `..cells.map(grid.cell)`)

Maybe we shouldn't care that much?

Reply:
You can add the table's span to the backtrace with `trace`.

Reply:
Is using `trace` necessary if the error already occurs from within the table's `Layout`? I was thinking here that, in order to use `trace`, `CellGrid::resolve` might have to take the `Tracing` as a parameter (alongside the `Span`) since it depends on the callee's context, though I guess that wouldn't be that bad.

Reply:
Okay, I added a `trace` call to grid and table in the meantime, producing the results below. We can change later if we implement a more general solution in the future.

![a `table.cell`-specific error now shows "help: error occurred in this call of function `table`".](https://github.com/typst/typst/assets/9021226/71624822-011c-416e-8c45-0abb72b8da9b)

![a `grid.cell`-specific error now shows "help: error occurred in this call of function `grid`".](https://github.com/typst/typst/assets/9021226/56af4be4-fef2-4af7-b1e4-21c8b21760bc)